fractional-NFT/project/token-distributor-contract/src/main.sw

```diff
@@ -0,0 +1,240 @@
+
+        require(token_distribution.owner.is_some() && token_distribution.owner.unwrap() == msg_sender().unwrap(), AccessError::NotFNftOwner);
+        require(token_distribution.state == DistributionState::Buyback || token_distribution.state == DistributionState::Started, AccessError::InvalidState);
+
+        // Update distribution state
+        token_distribution.state = DistributionState::Ended;
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+
+        // Send the remaining tokens to the fractionalized NFT contract and transfer ownership of the NFT to the owner
+        transfer(this_balance(fractional_nft), fractional_nft, Identity::ContractId(fractional_nft));
+        withdraw_fractional_nft(fractional_nft, token_distribution.owner.unwrap());
+
+        log(Closed { fractional_nft });
+    }
+
+    #[storage(read, write)]
+    fn purchase(amount: u64, fractional_nft: ContractId) {
+        let mut token_distribution = require_fractional_nft_exists(storage.token_distributions.get(fractional_nft));
+
+        // Make sure we are in purchasing state and if this is the first purchase change the state to distributing
+        require(token_distribution.state == DistributionState::Started || token_distribution.state == DistributionState::Distributed, AccessError::InvalidState);
+        if token_distribution.state == DistributionState::Started {
+            token_distribution.state = DistributionState::Distributed;
+        }
+
+        require(amount <= this_balance(fractional_nft), AssetError::NotEnoughTokensAvailable);
+        require(amount * token_distribution.token_price == msg_amount() && msg_asset_id() == token_distribution.external_asset, AssetError::InvalidAssetTransfer);
+
+        // Store the amount deposited and send the buyer some fractionalized NFT tokens
+        token_distribution.external_deposits += msg_amount();
+        storage.token_distributions.insert(fractional_nft, Option::Some(token_distribution));
+        transfer(amount, fractional_nft, msg_sender().unwrap());
+
+        log(Purchased {
+            amount,
+            buyer: msg_sender().unwrap(),
+            fractional_nft,
+        });
+    }
+
+    #[storage(read, write)]
+    fn purchase_ownership(
+        fractional_nft: ContractId,
+        owner: Option<Identity>,
+        reserve: Option<u64>,
+    ) {
+        let mut token_distribution = require_fractional_nft_exists(storage.token_distributions.get(fractional_nft));
+
+        require(token_distribution.reserve_price.is_some() && token_distribution.owner.is_some(), AccessError::NoReserveAvailable);
+        require(token_distribution.state != DistributionState::Ended, AccessError::InvalidState);
+        require(msg_asset_id() == token_distribution.external_asset && msg_amount() == token_distribution.reserve_price.unwrap(), AssetError::InvalidAssetTransfer);
```

Comment:
I'm uncertain about the logic here and if there is a problem. A buyer can purchase the NFT outright once users have already paid and have a stake in the NFT. The reserve seems to only be used for this. I haven't thought this through properly but I feel like there might be some edge case issue here.